CMakeLists.txt

```
@@ -248,6 +254,46 @@ else()
   )
 endif()
 
+# clang-tidy ---------------------------------------------------------------------------------------
+
+option(CS_RUN_CLANG_TIDY "Run clang-tidy with the compiler." Off)
+if(CS_RUN_CLANG_TIDY)
+  if(CMake_SOURCE_DIR STREQUAL CMake_BINARY_DIR)
+    message(FATAL_ERROR "CS_RUN_CLANG_TIDY requires an out-of-source build!")
+  endif()
+  find_program(CLANG_TIDY_COMMAND NAMES clang-tidy)
+  if(NOT CLANG_TIDY_COMMAND)
+    message(WARNING "CS_RUN_CLANG_TIDY is ON but clang-tidy is not found!")
+    set(CMAKE_CXX_CLANG_TIDY "" CACHE STRING "" FORCE)
+  else()
+    set(CS_CT_OPT "-quiet")
+    set(CS_CT_OPT "${CS_CT_OPT};-p;${CMAKE_BINARY_DIR}")
+
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-std=c++17")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-DGLM_ENABLE_EXPERIMENTAL")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-DGLM_FORCE_SWIZZLE")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_SOURCE_DIR}/src")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_BINARY_DIR}/src/cs-utils")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_BINARY_DIR}/src/cs-scene")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_BINARY_DIR}/src/cs-graphics")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_BINARY_DIR}/src/cs-core")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${CMAKE_BINARY_DIR}/src/cs-gui")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${COSMOSCOUT_EXTERNALS_DIR}/include")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${COSMOSCOUT_EXTERNALS_DIR}/include/cef")
+    set(CS_CT_OPT "${CS_CT_OPT};-extra-arg=-I${Boost_INCLUDE_DIRS}")
+
+    set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_COMMAND};${CS_CT_OPT}")
```

At least on Linux this is sufficient, as we create the `compile_commands.json` in `${CMAKE_BINARY_DIR}`. I have to test this on Windows though.

```suggestion
    set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_COMMAND};-quiet;-p;${CMAKE_BINARY_DIR}")
```

---

Did you get this working with the Visual Studio CMake generator? Or are you using Ninja?

---

On Windows it ran with MSBuild ("Visual Studio 16 2019" -A x64) and NMake ("CodeBlocks - NMake Makefiles"). Ninja is a little tricky with the MSVC toolchain.